File: 131.分割回文串.py
#
# @lc app=leetcode.cn id=131 lang=python3
#
# [131] 分割回文串
#

# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def isPalindrome(s: str) -> bool:
            # print("isPalindrome", s)
            i, j = 0, len(s) - 1
            while i < j:
                if s[i] != s[j]: return False
                i += 1
                j -= 1
            return True
        res = []
        path = []
        def helper(left):
            # print("path",path,"left", left)
            if left == len(s):
                res.append(path.copy())
                # res.append(path)
                return
            for right in range(left, len(s)):
                # print(s[left:right + 1])
                if isPalindrome(s[left:right + 1]):
                    path.append(s[left:right+1])
                    helper(right + 1)
                    path.pop()
            return
        helper(0)
        return res
# @lc code=end

File: 131.分割回文串.go
func isPalindrome(s string) bool {
    l, r := 0, len(s)-1
    for l < r {
        if s[l] != s[r] {
            return false
        }
        l++
        r--
    }
    return true
}
func partition(s string) [][]string {
    res := [][]string{}
    path := make([]string, 0)
    var helper func(int)
    helper = func(left int) {
        if left == len(s) {
            tmp := make([]string, len(path))
            copy(tmp, path)
            res = append(res, tmp)
            return
        }
        for right := left;  right < len(s); right++ {
            if isPalindrome(s[left:right+1]) {
                path = append(path, s[left:right+1])
                helper(right + 1)
                path = path[:len(path) - 1]
            }
        }
    }
    helper(0)
   // fmt.Println(path)
    return res
}
